### PR TITLE
Rianfowler/sc 50084/convert Dashboard into functional component 

### PR DESCRIPTION
**Pull request by rianfowler**
#### What this PR does / why we need it:
- Refactors Dashboard into functional component so we can use hooks in it 

#### Which issue(s) this PR fixes:

https://app.shortcut.com/replicated/story/50084/fix-multi-app-ux

#### Does this PR introduce a user-facing change?
none- refactor 

#### Does this PR require documentation?
none

### CODE REVIEW

File: web/src/components/apps/AppDetailPage.tsx [hunk 1]
```diff
@@ -251,12 +251,15 @@ function AppDetailPage(props: Props) {
     }
   };
 
-  const redeployVersion = async (upstreamSlug: string, version: Version) => {
+  const redeployVersion = async (
+    upstreamSlug: string,
+    version: Version | null
```
Comment (rianfowler) [inline web/src/components/apps/AppDetailPage.tsx:256]: small type change

File: web/src/components/apps/AppDetailPage.tsx [hunk 2]
```diff
@@ -435,11 +438,9 @@ function AppDetailPage(props: Props) {
                       app={selectedApp}
                       cluster={selectedApp.downstream?.cluster}
                       updateCallback={refetchData}
-                      onActiveInitSession={props.onActiveInitSession}
                       toggleIsBundleUploading={toggleIsBundleUploading}
                       makeCurrentVersion={makeCurrentRelease}
                       redeployVersion={redeployVersion}
-                      redeployVersionErrMsg={state.redeployVersionErrMsg}
```
Comment (rianfowler) [inline web/src/components/apps/AppDetailPage.tsx:442]: These are not used by the Dashboard component

File: web/src/features/Dashboard/components/Dashboard.tsx [hunk 3]
```diff
@@ -109,11 +110,13 @@ type State = {
...
-    super(props);
-
-    this.state = {
+const Dashboard = (props: Props) => {
+  const [state, setState] = useReducer(
+    (currentState: State, newState: Partial<State>) => ({
+      ...currentState,
+      ...newState,
+    }),
+    {
```
Comment (rianfowler) [inline web/src/features/Dashboard/components/Dashboard.tsx:119]: convert Dashboard into functional component

File: web/src/features/Dashboard/components/Dashboard.tsx [hunk 4]
```diff
@@ -423,21 +210,98 @@ class Dashboard extends Component<Props, State> {
...
+        });
+      });
+  };
+
+  useEffect(() => {
+    if (props.app) {
+      setWatchState(props.app);
+      getAppLicense(props.app);
+    }
+  }, [props.app]);
```
Comment (rianfowler) [inline web/src/features/Dashboard/components/Dashboard.tsx:278]: this replaces the `componentDidUpdate` with an effect hook that fires when the `props.app` updates

File: web/src/features/Dashboard/components/Dashboard.tsx [hunk 5]
```diff
@@ -721,311 +563,453 @@ class Dashboard extends Component<Props, State> {
...
+    if (app) {
+      setWatchState(app);
+      getAppLicense(app);
+    }
+    return () => {
+      state.updateChecker.stop();
+      state.getAppDashboardJob.stop();
+      state.fetchAppDownstreamJob.stop();
+    };
+  }, []);
```
Comment (rianfowler) [inline web/src/features/Dashboard/components/Dashboard.tsx:687]: replaces `componentDidMount` and `componentWillUnmount`

File: web/src/features/Dashboard/components/Dashboard.tsx [hunk 6]
```diff
@@ -721,311 +563,453 @@ class Dashboard extends Component<Props, State> {
...
+          )}
+        </div>
+      )}
+    </>
+  );
+};
 
 // TODO: fix withRouter type
 // eslint-disable-next-line
-export default withRouter(Dashboard) as any;
```
Comment (rianfowler) [inline web/src/features/Dashboard/components/Dashboard.tsx:1031]: replaced `withRouter` decorator with `useHistory` and `useRouteMatch` hooks

**Review by rianfowler — COMMENTED** [review]
most of the changes are from reordering the functions and removing `this.` Added notes to logic changes

File: web/src/components/apps/AppDetailPage.tsx (hunk 1 above)
Comment (rianfowler) [inline web/src/components/apps/AppDetailPage.tsx:256]: minor type change

File: web/src/features/Dashboard/components/Dashboard.tsx [hunk 7]
```diff
@@ -721,311 +563,453 @@ class Dashboard extends Component<Props, State> {
...
+// export default withRouter(Dashboard) as any;
```
Comment (alicenstar) [inline web/src/features/Dashboard/components/Dashboard.tsx:1013]: Can we remove this commented line now?

Comment (rianfowler) [reply web/src/features/Dashboard/components/Dashboard.tsx:1013]:
```suggestion
```

Comment (rianfowler) [reply web/src/features/Dashboard/components/Dashboard.tsx]: done

**Pull request merged**